pkg/authorization/authorizer/authorizer.go

```diff
@@ -0,0 +1,513 @@
+
+	return false, nil
+}
+
+func (a openshiftAuthorizationAttributes) verbMatches(rule authorizationapi.PolicyRule) bool {
+	verbMatches := false
+	verbMatches = verbMatches || contains(rule.Verbs, a.GetVerb())
+	verbMatches = verbMatches || contains(rule.Verbs, "*")
+
+	//check for negations that would force this match to false
+	verbMatches = verbMatches && !contains(rule.Verbs, "-"+a.GetVerb())
+	verbMatches = verbMatches && !contains(rule.Verbs, "-*")
+
+	return verbMatches
+}
+
+func (a openshiftAuthorizationAttributes) kindMatches(rule authorizationapi.PolicyRule) bool {
+	kindMatches := false
+	kindMatches = kindMatches || contains(rule.ResourceKinds, a.GetResourceKind())
+	kindMatches = kindMatches || contains(rule.ResourceKinds, "*")
+
+	//check for negations that would force this match to false
+	kindMatches = kindMatches && !contains(rule.ResourceKinds, "-"+a.GetResourceKind())
+	kindMatches = kindMatches && !contains(rule.ResourceKinds, "-*")
+
+	return kindMatches
+}
+
+func (a openshiftAuthorizationAttributes) GetUserInfo() authenticationapi.UserInfo {
+	return a.user
+}
+func (a openshiftAuthorizationAttributes) GetVerb() string {
+	return a.verb
+}
+func (a openshiftAuthorizationAttributes) GetResourceKind() string {
+	return a.resourceKind
+}
+func (a openshiftAuthorizationAttributes) GetNamespace() string {
+	return a.namespace
+}
+func (a openshiftAuthorizationAttributes) GetRequestAttributes() interface{} {
+	return a.requestAttributes
+}
+
+func (a *openshiftAuthorizationAttributeBuilder) GetAttributes(req *http.Request) (AuthorizationAttributes, error) {
+	verb, kind, namespace, _, err := VerbAndKindAndNamespace(req)
+	if err != nil {
+		return nil, err
+	}
+
+	userInterface, ok := a.requestsToUsers.Get(req)
+	if !ok {
+		return nil, errors.New("could not get user")
+	}
+	userInfo, ok := userInterface.(authenticationapi.UserInfo)
+	if !ok {
+		return nil, errors.New("wrong type returned for user")
+	}
+
+	return openshiftAuthorizationAttributes{
+		user:              userInfo,
+		verb:              verb,
+		resourceKind:      kind,
+		namespace:         namespace,
+		requestAttributes: nil,
+	}, nil
+}
+
+// TODO waiting on kube rebase
+// this section is copied from kube.  Need to modify kube to make this pluggable
+var specialVerbs = map[string]bool{
+	"proxy":    true,
+	"redirect": true,
+	"watch":    true,
+}
+
+func VerbAndKindAndNamespace(req *http.Request) (string, string, string, []string, error) {
+	parts := splitPath(req.URL.Path)
+
+	verb := ""
+	switch req.Method {
+	case "POST":
+		verb = "create"
+	case "GET":
+		verb = "get"
+	case "PUT":
+		verb = "update"
+	case "DELETE":
+		verb = "delete"
+	}
+
+	if parts[0] == "osapi" {
+		if len(parts) > 2 {
+			parts = parts[2:]
+		} else {
+			return "", "", "", nil, fmt.Errorf("Unable to determine kind and namespace from url, %v", req.URL)
+		}
+	}
+
+	// TODO tweak upstream to eliminate this copy
+	// handle input of form /api/{version}/* by adjusting special paths
```

Comment:
Note where this is copied from so we can see if there are changes we need to pick up. We want to use upstream logic ASAP, since keeping this up to date will be awful (namespace as param or path in different API versions, etc)